### PR TITLE
EN-44901: Primary table (aka this) alias and lateral joins

### PR DESCRIPTION
**Pull request by chitang**

### CODE REVIEW

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala [hunk 1]
```diff
@@ -96,10 +96,16 @@ object BinarySoQLAnalysisSqlizer extends Sqlizer[(BinaryTree[SoQLAnalysis[UserCo
...
-        val tableNamesSubTableNameReplace = tableNames + (TableName.PrimaryTable -> subTableName)
-        val primaryTableAlias = if (ra.joins.nonEmpty) Map((PrimaryTableAlias -> chainedTableAlias)) else Map.empty
+        val (primaryTableName, chainedTableAlias) = ra.from match {
+            case Some(tn@TableName(name, Some(alias))) if name == TableName.This =>
+              (tn.copy(alias = None), alias)
+            case _ =>
+              (TableName.PrimaryTable, "x1")
+          }
+        val subTableName = if (ra.from.isEmpty) "(%s) AS %s".format(lpsql.sql.head, chainedTableAlias)
+                           else "(%s)".format(lpsql.sql.head)
```
Comment (vickingan) [inline common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:106]: Is this the `this` case?

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala [hunk 2]
```diff
@@ -96,10 +96,16 @@ object BinarySoQLAnalysisSqlizer extends Sqlizer[(BinaryTree[SoQLAnalysis[UserCo
     analyses match {
       case PipeQuery(l, Leaf(ra)) =>
         val (lpsql, _) = sql(l, tableNames, allColumnReps, reqRowCount, rep, typeRep, setParams, ctx, escape, fromTableName)
-        val chainedTableAlias = "x1"
-        val subTableName = "(%s) AS %s".format(lpsql.sql.head, chainedTableAlias)
-        val tableNamesSubTableNameReplace = tableNames + (TableName.PrimaryTable -> subTableName)
-        val primaryTableAlias = if (ra.joins.nonEmpty) Map((PrimaryTableAlias -> chainedTableAlias)) else Map.empty
+        val (primaryTableName, chainedTableAlias) = ra.from match {
+            case Some(tn@TableName(name, Some(alias))) if name == TableName.This =>
+              (tn.copy(alias = None), alias)
```
Comment (vickingan) [inline common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:101]: Hm, so if the user tries to alias `this`, we ignore it?

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala [hunk 3]
```diff
@@ -96,10 +96,16 @@ object BinarySoQLAnalysisSqlizer extends Sqlizer[(BinaryTree[SoQLAnalysis[UserCo
...
+            case _ =>
+              (TableName.PrimaryTable, "x1")
+          }
+        val subTableName = if (ra.from.isEmpty) "(%s) AS %s".format(lpsql.sql.head, chainedTableAlias)
+                           else "(%s)".format(lpsql.sql.head)
+        val tableNamesSubTableNameReplace = tableNames + (primaryTableName -> subTableName)
+        val primaryTableAlias = if (ra.joins.nonEmpty) Map((PrimaryTableAlias -> chainedTableAlias)) else Map.empty // REVISIT
```
Comment (vickingan) [inline common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:108]: But if the user has aliased `this` during a join, we do use the provided alias? Looks like this was always defaulting to `x1` before this change--what does `x1` refer to?

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala (hunk 2 above)
Comment (chitang) [reply common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:101]: No, it is not ignored.  It is just the way we manage table aliases.

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala (hunk 1 above)
Comment (chitang) [reply common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:106]: the else block is the "this" case

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala (hunk 3 above)
Comment (chitang) [reply common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:108]: "x1" is the implicit (unused) alias we use for subquery (chained soql).  And it isn't new.

File: common-pg/src/main/scala/com/socrata/soql/analyzer/SoQLAnalyzerHelper.scala [hunk 4]
```diff
@@ -56,10 +56,17 @@ object SoQLAnalyzerHelper {
     analyses match {
       case PipeQuery(l, r) =>
         val nl = remapAnalyses(columnIdMapping, l)
-        val prev = nl.outputSchemaLeaf
+        val prev = nl.outputSchema.leaf
```
Comment (vickingan) [inline common-pg/src/main/scala/com/socrata/soql/analyzer/SoQLAnalyzerHelper.scala:59]: I'm guessing that this works the same as `nl.outputSchemaLeaf`?

**Review by vickingan — APPROVED** [review]
Left a few curious questions, LGTM 👍

File: common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala (hunk 3 above)
Comment (chitang) [reply common-pg/src/main/scala/com/socrata/pg/soql/SoQLAnalysisSqlizer.scala:108]: select c |> select c as c2 
will look like
select c as c3 from (select c) as x1

File: common-pg/src/main/scala/com/socrata/soql/analyzer/SoQLAnalyzerHelper.scala (hunk 4 above)
Comment (chitang) [reply common-pg/src/main/scala/com/socrata/soql/analyzer/SoQLAnalyzerHelper.scala:59]: yes, i refactor soql-references and that is an ongoing process.

**Pull request merged**